README
NAME
    Fuse::PDF - Filesystem embedded in a PDF document

SYNOPSIS
        use Fuse::PDF;
        my $fs = Fuse::PDF->new('my_doc.pdf');
        $fs->mount('/mnt/pdf');
        # blocks until the filesystem is unmounted

    See also the mount_pdf front-end.

LICENSE
    Copyright 2007 Chris Dolan, *cdolan@cpan.org*

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DESCRIPTION
    The Adobe Portable Document Format is an arbitrary collection of nodes
    which support a tree structure. Most of that data is oriented toward
    document rendering, but it's legal to add arbitrarily complex data
    virtually anywhere in the document structure. Adobe Illustrator does
    this to embed lots of metadata in it's "PDF-compatible" Illustrator
    document format.

    By deciding on a convention for representing a filesystem data and
    leveraging the FUSE (Filesystem in Userspace) library, we map filesystem
    calls to PDF edits.

BUGS AND CAVEATS
    Saving: No data is saved until you unmount the filesystem! Hopefully I
    can fix this in future releases. The saving is not yet atomic. That is,
    if you have a failure, the old PDF may be deleted before the new one is
    saved.

    Resources: The entire PDF is loaded into RAM in `new()'. If your
    filesystem grows too large, this will lead to obvious problems!

    Tests: There are no automated tests yet. That's next, I promise!

    Hangs: While FUSE is quite mature, I found it to be fairly easy to hang
    the filesystem while creating this package. I only needed to actually
    reboot once, but if that causes you concern you may wish to avoid FUSE
    in general.

    Operating systems: I've only tested this software with the Google build
    of MacFUSE 1.1.0 (PowerPC, 10.4, http://code.google.com/p/macfuse/).
    Notably, I have not tried the Linux implementation of FUSE. If you have
    other experiences to add, email me or (even better) post comments to
    http://annocpan.org/.

    Fuse.pm: As of this writing, the Fuse module (v0.09_01) fails all tests
    on Mac. The module actually *works* great, but the Makefile.PL and the
    tests are very Linux-centric. Hopefully that will improve as MacFUSE
    matures.

    PDF versions: This package relies on CAM::PDF to read and write PDFs.
    While that module supports all of the core PDF syntax, it's stricter
    than many other PDF implementations and may fail to open PDFs that, say,
    Acrobat or Preview.app can open. In particular, "Print to PDF" on Mac OS
    X 10.4 often generates bad PDFs.

    Threading: I've explicitly set the FUSE default to single-threaded mode,
    so performance may be terrible in some scenarios. I hope to add support
    for threaded Perl in a future release. Patches welcome (remove the
    `threaded => 0' line from this file and add locking to Fuse::PDF::FS).

    Unsupported: special files (named pipes, etc.), following symlinks out
    of the filesystem, permission enforcement, `chown', `flush', reading
    from unlinked filehandles.

    Performance: Don't expect this to be fast!

    find: I think I have the `nlinks' count wrong. According to the FUSE
    FAQ, this can confuse the 'find' command:
    http://fuse.sourceforge.net/wiki/index.php/FAQ#Why_doesnx27.t_find_work_
    on_my_filesystemx3f.

    Hard links: I have not yet implemented hard links. I'll consider
    implementing compressed streams at the same time.

METHODS
    $pkg->new($pdf_filename)
    $pkg->new($pdf_filename, $hash_of_options)
        Create a new filesystem instance. This method opens and parses the
        PDF document. If there is an error opening or parsing the PDF
        document, this will return `undef'.

        The options hash supports the following extra arguments:

        pdf_constructor
            An arrayref of extra arguments to pass to the CAM::PDF
            constructor. In particular, the first arguments are the owner
            and user password which can be used to open encrypted PDFs.

        save_filename
            The string representing the path where filesystem changes should
            be saved. By default this is the `$pdf_filename' passed to
            `new()'.

        compact
            A boolean indicating whether to discard old filesystem data
            saved via version infrastructure described in the PDF
            specification. Defaults to false. If left false, then the PDF
            will grow with every mount, but only by as much as you changed
            it. See `rewritepdf.pl --cleanse' from the CAM::PDF distribution
            to perform the compaction manually. See also `revertpdf.pl' to
            roll back to those older versions.

        fs_name
            Fuse::PDF can embed multiple filesystems in a single PDF
            distinguished by name. This string specifies which filesystem to
            use. It uses the Fuse::PDF::FS default if a name is not
            explicitly provided.

        revision
            A version number indicating which filesystem version to roll
            back to before mounting. Use `fs()' and the Fuse::PDF::FS API to
            learn what revisions are available in a PDF filesystem.

    $self->mount($mount_path)
    $self->mount($mount_path, $hash_of_fuse_options)
        Calls into Fuse to mount the filesystem to the specified mount
        point. On unmount, a new PDF will be saved with any filesystem
        changes.

        If the mount point does not exist, this package will try to create
        it as a directory via a simple `mkdir()'. If the `mkdir()' fails, or
        if the mount point exists but is not a directory, `mount()' will
        `croak()'.

        If the PDF has an existing filesystem which is incompatible with
        this version of the software, `mount()' will `croak()'.

        If the mount is successful, we establish callbacks and hand control
        to the FUSE library. FUSE blocks until the filesystem is unmounted.
        If this blocking is a problem for you, consider daemonizing the
        process like so:

            use Fuse::PDF;
            use Net::Server::Daemonize qw();
            my ($pdffilename, $mountdir) = @ARGV;
            my $fs = Fuse::PDF->new($pdffilename);
            if (Net::Server::Daemonize::safe_fork()) {
                exit 0; # parent process or failure
            }
            Net::Server::Daemonize::daemonize('www', 'www');
            $fs->mount($mountdir);

        The `mount()' method cleans up after itself sufficiently that you
        may call it again immediately after unmounting.

        The options hash is passed directly to `Fuse::main()'. See the Fuse
        documentation for the allowed keys. A simple example is:

            $fs->mount($mountdir, {debug => 1});

    $self->fs()
        Return a fresh copy of the Fuse::PDF::FS data structure representing
        this PDF. You should not try to manipulate this object while the
        filesystem is mounted. This module is not yet thread-safe!

SEE ALSO
    CAM::PDF

    Fuse

    mount_pdf

AUTHOR
    Chris Dolan, *cdolan@cpan.org*

CREDITS
    Thanks to the Madison Perl Mongers for thinking the idea was stupid
    enough that I was inspired to implement it!



Changes
# History of Perl distribution Fuse-PDF.  This change log is valid YAML.

0.07 :
     - Released: 2007 Nov 27
     - Summary: Expose embedded images under 'mount_pdf -A'

     - Feature: The 'mount_pdf -A' option shows embedded images as separate PDF files
     - Prerequisites: Added MIME::Base64 v3.0
     - Internals: Added Fuse::PDF::ImageTemplate, a skeleton for the image PDFs

0.06 :
     - Released: 2007 Nov 26
     - Summary: Implemented rename

     - Feature: Implemented fs_rename for Fuse::PDF::ContentFS (passthru mode) and Fuse::PDF::FS
     - Prerequisites: Test::Virtual::Filesystem v0.08 (for rename tests)
     - Internals: Better rename() symlink support in file search helper methods in Fuse::PDF::FS

     - Fix: Prevent uninitialized var warning in Fuse::PDF::fs()

0.05 :
     - Released: 2007 Nov 22
     - Summary: Filesystem for PDF itself

     - Feature: Filesystem to represent the PDF itself ('mount_pdf -A' or 'mount_pdf -f pdf')
       Credit: RGiersig@cpan.org
       URL: http://use.perl.org/comments.pl?sid=37564&cid=59050

     - Prerequisites: CAM::PDF v1.11
     - Prerequisites: Test::Virtual::Filesystem v0.06

     - Internals: Move POSIX:ENOATTR workaround to ErrnoHacks.pm
     - Internals: Move Fuse::PDF::FS::previous_revision implementation to CAM::PDF

0.04 :
     - Released: never (forgot to upload to CPAN!)
     - Summary: Track Test::Virtual::Filesystem v0.04 API changes

     - Fix: Test::Virtual::Filesystem renamed its test_* methods to enable_test_*
     - Internals: Fix Perl::Critic violations (v1.080)

     - Prerequisites: Test::Virtual::Filesystem v0.04

0.03 :
     - Released: 2007 Nov 16
     - Summary: xattr bug fixes

     - Fix: listxattr was broken - omitted errno
     - Fix: getxattr was broken - returned node instead of string
     - Workaround: Fuse.pm has the wrong XATTR_CREATE/REPLACE constants for Mac [rt.cpan.org #30790]
     - Workaround: POSIX.pm lacks ENOATTR (on Mac 10.4 at least)

     - Prerequisites: Test::Virtual::Filesystem v0.03

     - Testing: turned on Test::Virtual::Filesystem xattr tests
     - Testing: turned off several other Test::Virtual::Filesystem tests pre-emptively

     - Press: http://www.leyrer.priv.at/martin/2007/11/index.html#PDF-Filesystem

0.02 :
     - Released: 2007 Nov 15
     - Summary: End-user features; convert to Test::Virtual::Filesystem

     - Feature: delete an fs (mount_pdf --deletefs)
     - Feature: backup on save (mount_pdf -b)
     - Feature: show statistics (mount_pdf -i)
     - Feature: access older revisions (mount_pdf -r -k)
     - Feature: support multiple filesystems in a PDF (mount_pdf -f <name>)
     - Feature: pass options to FUSE (mount_pdf --fuseopts <opts>)

     - Prerequisites: Test::Virtual::Filesystem v0.02

     - Incompatibility: Increased file format number to 2
     - Incompatibility: Changed "mount_pdf --pass" to "mount_pdf --password"

0.01 :
     - Released: 2007 Nov 12
     - Summary: initial release

     - Caveats: see CAVEATS in Fuse::PDF documentation

     - Press: http://use.perl.org/article.pl?sid=07/11/15/223246


Build.PL
use Module::Build;

Module::Build->new(
   module_name => 'Fuse::PDF',
   dist_author => 'Chris Dolan <cdolan@cpan.org>',
   license => 'perl',
   requires => {
      'perl' => '5.8.0',    # authored against 5.8.6
      'Fuse' => '0.08',     # prefer 0.09 and beyond
      'CAM::PDF' => '1.11', # authored against 1.11
      'Readonly' => '1.00', # authored against 1.03
   },
   recommends => {
      # Needed for mount_pdf program
      'Getopt::Long' => '2.0',         # authored against 2.37    
      'Pod::Usage' => '1.0',           # authored against 1.35

      # Optional in tests, skipped if not present
      'Test::Pod' => 0,                # authored against 1.26
      'Test::Pod::Coverage' => 0,      # authored against 1.08
   },
   build_requires => {
      'Test::More' => 0.62,                   # authored against 0.72
      'Test::Class' => 0,                     # authored against 0.24
      'Test::Virtual::Filesystem' => '0.08',  # usually synchronized with t/fusepdf.t
   },
   script_files => [qw(
      bin/mount_pdf
   )],
   create_readme => 1,
   create_makefile_pl => 'traditional',
)->create_build_script;


lib/Fuse/PDF/ErrnoHacks.pm
#######################################################################
#      $URL: svn+ssh://equilibrious@equilibrious.net/home/equilibrious/svnrepos/chrisdolan/Fuse-PDF/lib/Fuse/PDF/ErrnoHacks.pm $
#     $Date: 2007-11-27 23:43:19 -0600 (Tue, 27 Nov 2007) $
#   $Author: equilibrious $
# $Revision: 724 $
########################################################################
package #
  Fuse::PDF::ErrnoHacks;

use warnings;
use strict;
use POSIX qw();
use Carp qw(carp);
use English qw(-no_match_vars);

our $VERSION = '0.07';

BEGIN {
   # ENOATTR isn't commonly defined in POSIX.pm.  Try to find it, or use a fallback value.
   if (!defined &POSIX::ENOATTR) {
      if (open my $fh, '<', '/usr/include/sys/errno.h') {
         my $content = do { local $INPUT_RECORD_SEPARATOR = undef; <$fh> };
         close $fh or carp 'Failed to close errno.h';
         if ($content =~ m/\#define \s+ ENOATTR \s+ (0x\d+|\d+)/xms) {
            my $errno = $1;
            if ($errno =~ m/0x(\d+)/xms) {
               $errno = hex $1;
            }
            *ENOATTR = sub { return $errno; };
         }
      }
      if (!defined *ENOATTR{CODE}) {
         *ENOATTR = sub { return POSIX::EIO(); };
      }
   }
}

sub import {
   if (!defined &POSIX::ENOATTR) {
      no strict 'refs';  ## no critic(TestingAndDebugging::ProhibitNoStrict)
      *{caller() . '::ENOATTR'} = \&ENOATTR;
   }
   return;
}

1;

__END__

=pod

=for stopwords POSIX.pm

=head1 NAME

Fuse::PDF::ErrnoHacks - Workaround for missing POSIX.pm errno values

=head1 SYNOPSIS

In each package that needs C<ENOATTR>:

   use Fuse::PDF::ErrnoHacks;

That's it.

=head1 DESCRIPTION

For reasons I do not understand, C<use POSIX ':errno_h'> is missing
some values on (at least) Mac OS X 10.4.  This module detects the
missing constant and tries to read from F</usr/include/sys/errno.h>
directly.  If it can't find the constant there either, it substitutes
C<EIO> instead.

=head1 GENERATED CONSTANTS

=over

=item ENOATTR()

=back

=head1 LICENSE

Copyright 2007 Chris Dolan, I<cdolan@cpan.org>

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SEE ALSO

L<POSIX>

=head1 AUTHOR

Chris Dolan, I<cdolan@cpan.org>

=cut

# Local Variables:
#   mode: perl
#   perl-indent-level: 3
#   cperl-indent-level: 3
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab :


lib/Fuse/PDF.pm
#######################################################################
#      $URL: svn+ssh://equilibrious@equilibrious.net/home/equilibrious/svnrepos/chrisdolan/Fuse-PDF/lib/Fuse/PDF.pm $
#     $Date: 2007-11-27 23:43:19 -0600 (Tue, 27 Nov 2007) $
#   $Author: equilibrious $
# $Revision: 724 $
########################################################################

package Fuse::PDF;

use warnings;
use strict;
use 5.008;

use English qw(-no_match_vars);
use Carp qw(croak carp);
use Fuse qw(:xattr);
use CAM::PDF;
use Fuse::PDF::ContentFS;
use Fuse::PDF::FS;

our $VERSION = '0.07';

{
   # Hack fix for Fuse 0.09 which has the wrong constants for Mac.  Get the
   # right values directly from the .h file, if available
   if (open my $fh, '<', '/usr/include/sys/xattr.h') {
      my $content = do { local $INPUT_RECORD_SEPARATOR = undef; <$fh> };
      close $fh or carp 'Failed to close xattr.h';
      my %consts = $content =~ m/\#define \s+ (XATTR_(?:CREATE|REPLACE)) \s+ (0x\d+|\d+)/gxms; ## no critic(RegularExpressions::ProhibitEscapedMetacharacters) -- Perl::Critic v1.08 bug...
      for my $name (keys %consts) {
         my $value = $consts{$name} || next;
         if ($value =~ m/0x(\d+)/xms) {
            $value = hex $1;
         }
         no warnings 'redefine'; ## no critic(TestingAndDebugging::ProhibitNoWarnings)
         no strict 'refs';       ## no critic(TestingAndDebugging::ProhibitNoStrict)
         *{$name} = sub { return $value; };
      }
   }
}


sub new {
   my ($pkg, $filename, $extra) = @_;
   croak 'No PDF filename specified' if !defined $filename;
   croak 'No PDF file found'         if !-e $filename;
   $extra ||= {};
   my $pdf = CAM::PDF->new($filename, @{ $extra->{pdf_constructor} || [] });
   return if !$pdf;
   my $self = bless {
      save_filename => $filename,
      compact       => 0,
      fs_name       => undef,
      %{$extra},
      pdf_mtime => -e $filename ? (stat $filename)[9] : $BASETIME,  ## no critic(MagicNumber)
      pdf => $pdf,
   }, $pkg;
   return $self;
}

sub fs {
   my ($self) = @_;

   my $fs_class = $self->{fs_name} && $self->{fs_name} eq 'pdf'
       ? 'Fuse::PDF::ContentFS' : 'Fuse::PDF::FS';

   my $fs = $fs_class->new({
      pdf_mtime => $self->{pdf_mtime},
      pdf => $self->{pdf},
      fs_name => $self->{fs_name},
   }) or croak 'Failed to open the PDF or create a filesystem inside';

   if ($self->{revision}) {
      my @revisions = $fs->all_revisions;
      if ($self->{revision} =~ m/\D/xms && $self->{revision} <= @revisions) {
         $fs = $revisions[@revisions - $self->{revision}];
      } else {
         croak 'Invalid filesystem revision: ' . $self->{revision};
      }
   }

   return $fs;
}

sub mount {  ## no critic(Subroutines::ProhibitExcessComplexity)
   my ($self, $mountdir, $options) = @_;
   $options ||= {};

   my $fs = $self->fs;
   $fs->autosave_filename($self->{save_filename});
   $fs->compact($self->{compact});
   $fs->backup($self->{backup});

   # perform mount
   if (!-d $mountdir) {
      mkdir $mountdir;
      if (!-d $mountdir) {
         croak 'Failed to create mount directory: ' . $OS_ERROR;
      }
   }

   Fuse::main(
      mountpoint => $mountdir,

      $fs->can('fs_getattr') ? (
         getattr => sub {
            my ($file) = @_;
            return $fs->fs_getattr($file);
         }
      ) : (),
      $fs->can('fs_readlink') ? (
         readlink => sub {
            my ($file) = @_;
            return $fs->fs_readlink($file);
         }
      ) : (),
      $fs->can('fs_getdir') ? (
         getdir => sub {
            my ($file) = @_;
            return $fs->fs_getdir($file);
         }
      ) : (),
      $fs->can('fs_mknod') ? (
         mknod => sub {
            my ($file, $modes, $dev) = @_;
            return $fs->fs_mknod($file, $modes, $dev);
         }
      ) : (),
      $fs->can('fs_mkdir') ? (
         mkdir => sub {
            my ($file, $perms) = @_;
            return $fs->fs_mkdir($file, $perms);
         }
      ) : (),
      $fs->can('fs_unlink') ? (
         unlink => sub {
            my ($file) = @_;
            return $fs->fs_unlink($file);
         }
      ) : (),
      $fs->can('fs_rmdir') ? (
         rmdir => sub {
            my ($file) = @_;
            return $fs->fs_rmdir($file);
         }
      ) : (),
      $fs->can('fs_symlink') ? (
         symlink => sub {
            my ($link, $file) = @_;
            return $fs->fs_symlink($link, $file);
         }
      ) : (),
      $fs->can('fs_rename') ? (
         rename => sub {
            my ($oldfile, $file) = @_;
            return $fs->fs_rename($oldfile, $file);
         }
      ) : (),
      $fs->can('fs_link') ? (
         link => sub {
            my ($srcfile, $file) = @_;
            return $fs->fs_link($srcfile, $file);
         }
      ) : (),
      $fs->can('fs_chmod') ? (
         chmod => sub {
            my ($file, $perms) = @_;
            return $fs->fs_chmod($file, $perms);
         }
      ) : (),
      $fs->can('fs_chown') ? (
         chown => sub {
            my ($file, $uid, $gid) = @_;
            return $fs->fs_chown($file, $uid, $gid);
         }
      ) : (),
      $fs->can('fs_truncate') ? (
         truncate => sub {
            my ($file, $length) = @_;
            return $fs->fs_truncate($file, $length);
         }
      ) : (),
      $fs->can('fs_utime') ? (
         utime => sub {
            my ($file, $atime, $utime) = @_;
            return $fs->fs_utime($file, $atime, $utime);
         }
      ) : (),
      $fs->can('fs_open') ? (
         open => sub {
            my ($file, $mode) = @_;
            return $fs->fs_open($file, $mode);
         }
      ) : (),
      $fs->can('fs_read') ? (
         read => sub {
            my ($file, $size, $offset) = @_;
            return $fs->fs_read($file, $size, $offset);
         }
      ) : (),
      $fs->can('fs_write') ? (
         write => sub {
            my ($file, $str, $offset) = @_;
            return $fs->fs_write($file, $str, $offset);
         }
      ) : (),
      $fs->can('fs_statfs') ? (
         statfs => sub {
            return $fs->fs_statfs();
         }
      ) : (),
      $fs->can('fs_flush') ? (
         flush => sub {
            my ($file) = @_;
            return $fs->fs_flush($file);
         }
      ) : (),
      $fs->can('fs_release') ? (
         release => sub {
            my ($file, $mode) = @_;
            return $fs->fs_release($file, $mode);
         }
      ) : (),
      $fs->can('fs_fsync') ? (
         fsync => sub {
            my ($file, $flags) = @_;
            return $fs->fs_fsync($file, $flags);
         }
      ) : (),
      $fs->can('fs_setxattr') ? (
         setxattr => sub {
            my ($file, $key, $value, $flags) = @_;
            my %flags = (
               create  => $flags & XATTR_CREATE(),
               replace => $flags & XATTR_REPLACE(),
                );
            return $fs->fs_setxattr($file, $key, $value, \%flags);
         }
      ) : (),
      $fs->can('fs_getxattr') ? (
         getxattr => sub {
            my ($file, $key) = @_;
            return $fs->fs_getxattr($file, $key);
         }
      ) : (),
      $fs->can('fs_listxattr') ? (
         listxattr => sub {
            my ($file, $key) = @_;
            return $fs->fs_listxattr($file);
         }
      ) : (),
      $fs->can('fs_removexattr') ? (
         removexattr => sub {
            my ($file, $key) = @_;
            return $fs->fs_removexattr($file, $key);
         }
      ) : (),

      threaded => 0,
      %{$options},
   );
   return;
}

1;

__END__

=pod

=head1 NAME

Fuse::PDF - Filesystem embedded in a PDF document

=head1 SYNOPSIS

    use Fuse::PDF;
    my $fs = Fuse::PDF->new('my_doc.pdf');
    $fs->mount('/mnt/pdf');
    # blocks until the filesystem is unmounted

See also the F<mount_pdf> front-end.

=head1 LICENSE

Copyright 2007 Chris Dolan, I<cdolan@cpan.org>

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 DESCRIPTION

The Adobe Portable Document Format is an arbitrary collection of nodes
which support a tree structure.  Most of that data is oriented toward
document rendering, but it's legal to add arbitrarily complex data
virtually anywhere in the document structure.  Adobe Illustrator does
this to embed lots of metadata in it's "PDF-compatible" Illustrator
document format.

By deciding on a convention for representing a filesystem data and
leveraging the FUSE (Filesystem in Userspace) library, we map
filesystem calls to PDF edits.

=head1 BUGS AND CAVEATS

B<Saving:> B<No data is saved until you unmount the filesystem!>
Hopefully I can fix this in future releases.  The saving is not yet
atomic.  That is, if you have a failure, the old PDF may be
deleted before the new one is saved.

B<Resources:> The B<entire> PDF is loaded into RAM in C<new()>.  If
your filesystem grows too large, this will lead to obvious problems!

B<Tests:> There are no automated tests yet.  That's next, I promise!

B<Hangs:> While FUSE is quite mature, I found it to be fairly easy to hang the
filesystem while creating this package.  I only needed to actually
reboot once, but if that causes you concern you may wish to avoid FUSE
in general.

B<Operating systems:> I've only tested this software with the Google build of MacFUSE 1.1.0
(PowerPC, 10.4, L<http://code.google.com/p/macfuse/>).  Notably, I
have not tried the Linux implementation of FUSE.  If you have other
experiences to add, email me or (even better) post comments to
L<http://annocpan.org/>.

B<Fuse.pm:> As of this writing, the L<Fuse> module (v0.09_01) fails all tests on Mac.  The
module actually I<works> great, but the F<Makefile.PL> and the tests are very
Linux-centric.  Hopefully that will improve as MacFUSE matures.

B<PDF versions:> This package relies on L<CAM::PDF> to read and write PDFs.  While that
module supports all of the core PDF syntax, it's stricter than many other
PDF implementations and may fail to open PDFs that, say, Acrobat or
Preview.app can open.  In particular, "Print to PDF" on Mac OS X 10.4
often generates bad PDFs.

B<Threading:> I've explicitly set the FUSE default to single-threaded
mode, so performance may be terrible in some scenarios.  I hope to add
support for threaded Perl in a future release.  Patches welcome
(remove the C<threaded =E<gt> 0> line from this file and add locking
to L<Fuse::PDF::FS>).

B<Unsupported:> special files (named pipes, etc.), following symlinks
out of the filesystem, permission enforcement, C<chown>, C<flush>,
reading from unlinked filehandles.

B<Performance:> Don't expect this to be fast!

B<find:> I think I have the C<nlinks> count wrong.  According to the FUSE FAQ,
this can confuse the 'find' command:
L<http://fuse.sourceforge.net/wiki/index.php/FAQ#Why_doesnx27.t_find_work_on_my_filesystemx3f.>

B<Hard links:> I have not yet implemented hard links.  I'll consider
implementing compressed streams at the same time.

=head1 METHODS

=over

=item $pkg->new($pdf_filename)

=item $pkg->new($pdf_filename, $hash_of_options)

Create a new filesystem instance.  This method opens and parses the
PDF document.  If there is an error opening or parsing the PDF
document, this will return C<undef>.

The options hash supports the following extra arguments:

=over

=item pdf_constructor

An arrayref of extra arguments to pass to the L<CAM::PDF> constructor.
In particular, the first arguments are the owner and user password
which can be used to open encrypted PDFs.

=item save_filename

The string representing the path where filesystem changes should be
saved.  By default this is the C<$pdf_filename> passed to C<new()>.

=item compact

A boolean indicating whether to discard old filesystem data saved via
version infrastructure described in the PDF specification.
Defaults to false.  If left false, then the PDF will grow with every
mount, but only by as much as you changed it.  See C<rewritepdf.pl
--cleanse> from the L<CAM::PDF> distribution to perform the compaction
manually.  See also C<revertpdf.pl> to roll back to those older
versions.

=item fs_name

Fuse::PDF can embed multiple filesystems in a single PDF distinguished
by name.  This string specifies which filesystem to use.  It uses the
Fuse::PDF::FS default if a name is not explicitly provided.

=item revision

A version number indicating which filesystem version to roll back to
before mounting.  Use C<fs()> and the L<Fuse::PDF::FS> API to learn
what revisions are available in a PDF filesystem.

=back

=item $self->mount($mount_path)

=item $self->mount($mount_path, $hash_of_fuse_options)

Calls into Fuse to mount the filesystem to the specified mount point.
On unmount, a new PDF will be saved with any filesystem changes.

If the mount point does not exist, this package will try to create it
as a directory via a simple C<mkdir()>.  If the C<mkdir()> fails, or
if the mount point exists but is not a directory, C<mount()> will
C<croak()>.

If the PDF has an existing filesystem which is incompatible with
this version of the software, C<mount()> will C<croak()>.

If the mount is successful, we establish callbacks and hand control to
the FUSE library.  FUSE blocks until the filesystem is unmounted.  If
this blocking is a problem for you, consider daemonizing the process
like so:

    use Fuse::PDF;
    use Net::Server::Daemonize qw();
    my ($pdffilename, $mountdir) = @ARGV;
    my $fs = Fuse::PDF->new($pdffilename);
    if (Net::Server::Daemonize::safe_fork()) {
        exit 0; # parent process or failure
    }
    Net::Server::Daemonize::daemonize('www', 'www');
    $fs->mount($mountdir);

The C<mount()> method cleans up after itself sufficiently that you may
call it again immediately after unmounting.

The options hash is passed directly to C<Fuse::main()>.  See the
L<Fuse> documentation for the allowed keys.  A simple example is:

    $fs->mount($mountdir, {debug => 1});

=item $self->fs()

Return a fresh copy of the L<Fuse::PDF::FS> data structure representing this
PDF.  You should not try to manipulate this object while the filesystem is
mounted.  This module is not yet thread-safe!

=back

=head1 SEE ALSO

L<CAM::PDF>

L<Fuse>

F<mount_pdf>

=head1 AUTHOR

Chris Dolan, I<cdolan@cpan.org>

=head1 CREDITS

Thanks to the Madison Perl Mongers for thinking the idea was stupid
enough that I was inspired to implement it!

=cut

# Local Variables:
#   mode: perl
#   perl-indent-level: 3
#   cperl-indent-level: 3
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab :
